### ISSUE TITLE
Scala grammar testability issues and possible fix

### ISSUE THREAD

**Issue by Marti2203** [antlr/grammars-v4#1983]
Hello,
Two weeks ago I started to play around with the Scala Grammar. 
There were a couple of issues - the grammar does not support XML patterns(will have to add that) and the current implementation did not work(it passed tests,but it was not made to try to reach EOF so the parser stopped at the first moment it could not be sure about the input, which was just before the opening brace I suppose).

After a lot of tweaking I have made it working, but with one caveat - I had to make a custom TokenStream as the parser by specification needs to be contextually aware to know whether to emit newline tokens or not (this is just for Scala 2, probably for Scala 3 another rework will be needed as it is becoming an indentation based language). I am not very well versed in the current testing facilities and would like to know what is the proper approach if I were to try making a Pull Request for this?

Probably more work would be needed or something better could be suggested as a replacement such that the TokenStream need not exist.

**Comment by studentmain** [comment]
For now, I'm putting all necessary tools (except pom.xml) here, and hope next developer remember to run it. It's not intergrated into CI, and only java runtime is tested, but at least it works on our local machine.

If a new test framework is too far, maybe we can write test scripts when current test framework not works, and run it in CI.

**Comment by Marti2203** [comment]
But this doe not make it possible to properly test the grammar through the CI/CD, I would like to have that.

**Comment by aryx** [comment]
@Marti2203 is your tokenstream available somewhere? from what you said it seems like the Scala ANTLR grammar can not parse correctly Scala files right?

**Comment by Marti2203** [comment]
I can send it privately. I am still not okay with including it in the repo.

**Comment by Marti2203** [comment]
@kaby76 , you are currently developping the testability suite for this repo right?

**Comment by kaby76** [comment]
@Marti2203 Yes, I have been writing the trgen testing tool, but it's generally driven by the existence of the maven pom file and templates in grammars-g4/_scripts/templates/. We can still test it with some custom templates for code generation. If one can write a Java driver that does work, it can be plugged into testing as a separate step. -Ken

**Comment by Marti2203** [comment]
@kaby76  Okay, that is good to know. I am still working on making the grammar able to parse the Scala repo itself and another Scala project that is  popular - XML is almost fully parseable, even with nestedness of interpolated strings. It is not possible to make it work without the custom token stream, not sure if I can integrate it into the lexer, will try that later on.

**Comment by Vladimir-Chan** [comment]
There were lots of issues in the Scala Grammar, but this is not author's fault. In fact, as mentioned in grammar file, it derived from official grammar document. In my opinion, supporting xml or generating NL tokens correctly is insignificant. What we should do is to modifier the grammar to fix the issue - early stopping before reaching EOF just like the following:
`object HelloWorld {
   def main(args: Array[String]) {
	  var a = 3
      var b = 4
	  var c = a * b
	   println(c)
   }
}`
In this example, * was not recognized correctly, so that the node classTemplateOpt of parse tree has no child.

**Comment by Vladimir-Chan** [comment]
There were lots of issues in the Scala Grammar, but this is not author's fault. In fact, as mentioned in grammar file, it derived from official grammar document. In my opinion, supporting xml or generating NL tokens correctly is insignificant. What we should do is to modifier the grammar to fix the issue - early stopping before reaching EOF just like the following:
```
object HelloWorld {
    def main(args: Array[String]) {
        var a = 3
        var b = 4
        var c = a * b
        println(c)
    }
}
```
In this example, * was not recognized correctly, so that the node classTemplateOpt of parse tree has no child.

**Comment by Marti2203** [comment]
Hi, I understand that it is not the author's fault, I never blamed them. The official grammar document is really sub par, I have seen a lot of places in the official parser code where they use Scala code that is invalid per the grammar. I am currently working it to parse 99% of files and I am close to done after a long time.

**Comment by Vladimir-Chan** [comment]
I can't agree more. Python's official grammar document is much better than it. There have been no problems during I use antlr4 to parse python files.

**Comment by kaby76** [comment]
@Marti2203 Are you deriving your grammar from the [so-called "language spec"](https://scala-lang.org/files/archive/spec/2.13/13-syntax-summary.html) or scraping the grammar from the sources, either a hand-written parser (which I can't find anywhere in https://github.com/scala/scala), or a grammar presumably in a [combinator implementation](https://www.scala-lang.org/api/2.12.3/scala-parser-combinators/scala/util/parsing/combinator/index.html), which I also cannot find anywhere?

**Comment by aryx** [comment]
The LL(1) parser is here: https://github.com/scala/scala/tree/2.13.x/src/compiler/scala/tools/nsc/ast/parser

**Comment by aryx** [comment]
There's also a combinator-based Scala parser here: https://github.com/com-lihaoyi/fastparse/tree/master/scalaparse/src/scalaparse but it's quite tricky to understand because the combinary library use advanced Scala features (e.g., implicits).

**Comment by aryx** [comment]
I also tried to write a parser based on the language-spec but abandonned and opted to just mimic the recursive descent parser in the scala compiler source (my implementation is here: https://github.com/returntocorp/pfff/blob/develop/lang_scala/parsing/Parser_scala_recursive_descent.ml )

**Comment by Marti2203** [comment]
@kaby76  @aryx I am using the "language spec" and just testing the build against the scala compiler itself and some extra test files.

**Comment by aryx** [comment]
A non-trivial unusual thing is Scala is how to handle newlines; In most languages they are skipped. In scala they must sometimes be maintained, sometimes skipped, depending on non-trivial conditions on the tokens around.

**Comment by Marti2203** [comment]
Yeah, I am almost done with it - currently catching obvious "edge cases". I am diverging a little from the grammar. It won't be perfect but parsing 90% of the files is better.

**Comment by kaby76** [comment]
The parser for Scala compiler is at https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/ast/parser/Parsers.scala , and it contains doc comments that I suspect are used to generate the "language spec" Chapter 13 grammar at https://scala-lang.org/files/archive/spec/2.13/13-syntax-summary.html . And, I've been around the block a few times (since '84) to know that comments are rarely kept up to date. At some point, I will scrape the grammar from the code.

**Comment by Marti2203** [comment]
Yeah... With the little experience I have, I expect the same thing, but it is close enough to parse.

**Comment by Vladimir-Chan** [comment]
Okay, we can check the last token in input stream to ensure correct parse just like this
```
input_stream = FileStream(...)
lexer = Lexer(input_stream)
token_stream = CommonTokenStream(lexer)
parser = Parser(token_stream)
parser._errHandler = BailErrorStrategy()
parse_tree = parser.compilationUnit()
assert token_stream.LT(1).type == -1
```

**Comment by Vladimir-Chan** [comment]
Okay, we can check the last token in input stream to ensure correct parse just like this
```Python
input_stream = FileStream(...)
lexer = Lexer(input_stream)
token_stream = CommonTokenStream(lexer)
parser = Parser(token_stream)
parser._errHandler = BailErrorStrategy()
parse_tree = parser.compilationUnit()
assert token_stream.LT(1).type == -1
```
But there is still a issue that infixExpr node will be parsed incorrectly when parsing following:
```Scala
object HelloWorld {
    def main(args: Array[String]) {
        var a = 3
	var b = 4
	a += b
	println(a)
    }
}
```
`a` in `a += b` will be recognized as `Id`, and preceding token `4` will be recognized as the first infixExpr, following tokens will be recognized as the second infixExpr in production `infixExpr : infixExpr Id NL? infixExpr`. It is obviously incorrect.

**Comment by Marti2203** [comment]
The current solution is a token stream that keeps track of the state of the current NL token -is it between braces or parenthesis. Let me finally make a PR after so long so that people can see what I am doing as I feel like it looks like I am doing vaporware. 😅
